lib/matplotlib/axes/_axes.py

```diff
@@ -2604,6 +2605,123 @@ def barh(self, y, width, height=0.8, left=None, *, align="center",
                            align=align, **kwargs)
         return patches
 
+    def bar_label(self, container, captions=[], *, fmt="%g", position="edge",
+                  padding=0, shifting=0, **kwargs):
+        """
+        Label a bar plot.
+
+        Adds labels to bars in the given `.BarContainer`.
+        You may need to adjust the axis limits to fit the labels.
+
+        Parameters
+        ----------
+        container : `.BarContainer`
+            Container with all the bars and optionally errorbars.
+
+        captions : array-like, optional
+            A list of label texts, that should be displayed. If not given, the
+            label texts will be the data values formatted with *fmt*.
+
+        fmt : str, default: '%g'
+            A format string for the label.
+
+        position : {'edge', 'center'}, default: 'edge'
+            Position of the label relative to the bar:
+
+            - 'edge': Placed at edge, the cumulative values will be shown.
+            - 'center': Placed at center, the individual values will be shown.
+
+        padding : float, default: 0
```

Comment:
Instead of "longitudinal" and "lateral" how about "{parallel, perpendicular} to the direction of the bar" ?

Reply:
I'd suggest: 
```python
`padding` : float, default: 0
      Distance of label from end of bar
`offset` : float default: 0
     Distance label is moved relative to the centre line of the bar (positive up/right)
```

Reply:
currently the `padding` is still effective when position is 'center' and its direction will be the same as the bar points to, how could we describe this behavior? (or we disable it for 'center'?)

![68781801-46f9c080-0673-11ea-8b76-fa4154bdd36ff](https://user-images.githubusercontent.com/26521816/68782000-963ff100-0673-11ea-9c80-fb9787f89a24.png)

Reply:
Do we need the shift perpendicular to the bar?

1. I can't imagine a plot that wants this shift. Anybody has a good example?
2. As it is implemented no, it's difficult to use because the label is centered in that direction. And offsets relative to center are somewhat messy (neither the centers nor the boundaries of the labels are very distinct reference points).

Reply:
Admittedly, perpendicular shifting is a bit over-design, I would vote for remove it because I can't think of a justified / common use case either.

Reply:
Shift by (size of bar)/2 to place the label centered, but over the bar?

Reply:
place the label centered, but over the bar does not look good

![image](https://user-images.githubusercontent.com/26521816/93023522-23e8c600-f622-11ea-9fc2-021480d29e33.png)

Reply:
That doesn't mean people won't want to do it. It would be fine with a slightly smaller font, or larger gap between bars.

Reply:
maybe we need more opinions on the "shifting" topic? because previous comment suggests to remove it and current comment suggests adding it back, i am so confused  :confused:

Reply:
My opinion is to leave it off for now.  Folks can still manually place the text if they don't like the helper.  We can always add it later if its requested.  

OTOH, don't be confused.  Different folks have different opinions.  @timhoffm is now API lead so if we don't get a consensus in a day or so, we can ask him for an explicit ruling :wink:

Reply:
got it & thanks for the reply!

Reply:
I agree with @jklymak. Leave perpendicular shifts out for now.

Reply:
@immaxchen  is this taken care of now?  (being lazy and not totally refreshing myself on this).  If so, I think you are just waiting for @QuLogic  to unblock.  If you have addressed all his concerns this should be very close...

Reply:
I think just @QuLogic need some time to think about if adding `datavalues` into `BarContainer` is a good idea, other suggestions should all be already addressed